app/services/user_service.rb

```diff
@@ -4,9 +4,27 @@ def self.delete_spam(user)
     user.destroy
   end
 
+  def self.deactivate(user:, actor:, params:)
+    actor.ability.authorize! :deactivate, user
+    UserDeactivationResponse.create(user: user, body: params[:deactivation_response])
+    user.deactivate!
+    user
+  end
+
   def self.update(user:, actor:, params:)
     actor.ability.authorize! :update, user
-    user.update! params
+    user.update params
+    user
+  end
+
+  def self.change_password(user:, actor:, params:)
+    actor.ability.authorize! :update, user
+    user.errors.add :current_password, I18n.t(:"error.current_password_did_not_match") unless user.valid_password? params.delete(:current_password)
```

Comment:
I feel like this is some of the crappier code I've written recently.. I had another implementation where I had custom validators on the User model for this stuff, but that involved hanging an extra attr_accessor off of the user model for `current_password`, as well as two whole new custom validation classes (additional complexity in user.rb = :angry:) for what amounts to these two lines of validation code.